### DISCUSSION TITLE
OceanBase Community ByLaws

### DISCUSSION THREAD

**Discussion by longdafeng** [oceanbase/oceanbase#891] [Classroom]
## Roles and Responsibilities
OceanBase Community define a set of roles with associated rights and responsibilities. These roles govern what tasks an individual may perform within the project. Please refer to [https://open.oceanbase.com/community/organization](https://open.oceanbase.com/community/organization) for the roles. 
### 
## Voting
Decisions regarding the projects under https://github.com/OceanBase are made by votes on the primary project development ([https://github.com/oceanbase/oceanbase/discussions](https://github.com/oceanbase/oceanbase/discussions)). Where necessary, PMC voting may take place on the private  SIG mailing list. Votes are clearly indicated by subject line starting with [VOTE]. Votes may contain multiple items for approval and these should be clearly separated. Voting is carried out by replying to the vote mail or discussion. Voting may take four flavors:

| Vote | Meaning |
| --- | --- |
| +1 | 'Yes,' 'Agree,' or 'the action should be performed.' |
| +0 | Neutral about the proposed action. |
| -0 | Mildly negative, but not enough so to want to block it. |
| -1 | This is a negative vote. On issues where consensus is required, this vote counts as a veto. All vetoes must contain an explanation of why the veto is appropriate. Vetoes with no explanation are void. It may also be appropriate for a -1 vote to include an alternative course of action. |

All participants in the OceanBase project are encouraged to show their agreement with or against a particular action by voting. For technical decisions, only the votes of PMCs  or TOCs are binding. Non-binding votes are still useful for those with binding votes to understand the perception of an action in the wider OceanBase community. For PMC decisions, only the votes of PMC or TOC members are binding.
Voting can also be applied to changes already made to the OceanBase codebase. These typically take the form of a veto (-1) in reply to the commit message sent when the commit is made. Note that this should be a rare occurrence. All efforts should be made to discuss issues when they are still patches before the code is committed.
Only PMC or TOC members have binding votes.
## Approvals
These are the types of approvals that can be sought. Different actions require different types of approvals

| Approval Type | Criteria |
| --- | --- |
| Consensus Approval | Consensus approval requires 3 binding +1 votes and no binding vetoes. |
| Majority Approval | Majority approval requires at least 3 binding +1 votes and more +3 votes than -1 votes. |
| Lazy Consensus | Lazy consensus requires no -1 votes ('silence gives assent'). |
| 2/3 Majority | 2/3 majority votes requires at least 3 votes and twice as many +1 votes as -1 votes. |

### Vetoes
A valid, binding veto cannot be overruled. If a veto is cast, it must be accompanied by a valid reason explaining the reasons for the veto. The validity of a veto, if challenged, can be confirmed by anyone who has a binding vote. This does not necessarily signify agreement with the veto - merely that the veto is valid.
If you disagree with a valid veto, you must lobby the person casting the veto to withdraw their veto. If a veto is not withdrawn, any action that has been vetoed must be reversed in a timely manner.
## Actions
This section describes the various actions which are undertaken within the project, the corresponding approval required for that action and those who have binding votes over the action.

| Actions | Description | Approval | Binding Votes | Minimum Length | T |
| --- | --- | --- | --- | --- | --- |
| Code Change | A change made to a source code of the project and committed by a Committer. | A minimum of one +1 from a PMC other than the one who authored the patch, and no -1s. The code can be committed after the first +1. If a -1 is received to the patch within 7 days after the patch was posted, it may be reverted immediately if it was already merged. | Active PMC members | 1 day from initial patch (**Note:** Committers should consider allowing more time for review based on the complexity and/or impact of the patch in question.) | Github/Gitee/Gitcode/GitLink pull  |
| Non-Code Change | A change made to a repository of the project and committed by a Committer. This includes documentation, website content, etc., but not source code, unless only comments are being modified. | Lazy Consensus | Active committer members | At the discression of the Committer | Github/Gitee/Gitcode/GitLink pull  |
| Product Release | A vote is required to accept a proposed release as an official release of the project. Any Committer may call for a release vote at any point in time. | Majority Approval | Active PMC members | 3 days | [https://github.com/oceanbase/oceanbase/discussions](https://github.com/oceanbase/oceanbase/discussions) |
| Adoption of New Codebase | When the codebase for an existing, released product is to be replaced with an alternative codebase. If such a vote fails to gain approval, the existing code base will continue. This also covers the creation of new sub-projects and submodules within the project as well as merging of feature branches. | 2/3 Majority | Active PMC members | 4 days | [https://github.com/oceanbase/oceanbase/discussions](https://github.com/oceanbase/oceanbase/discussions) |
| New Committer | When a new Committer is proposed for the project. | Consensus Approval | Active PMC members | 3 days | Every SIG private email list |
| New PMC Member | When a member is proposed for the PMC. | Consensus Approval | Active PMC members | 3 days | Every SIG private email list |
| Emeritus PMC Member re-instatement | When an emeritus PMC member requests to be re-instated as an active PMC member. | Consensus Approval | Active PMC members | 4 days | Every SIG private email list |
| Emeritus Committer re-instatement | When an emeritus Committer requests to be re-instated as an active Committer. | Consensus Approval | Active PMC members | 4 days | Every SIG private email list |
| Committer Removal | When removal of commit privileges is sought. Note: Such actions will be triggered by TOC | 2/3 Majority | Active PMC members (excluding the Committer in question if a member of the PMC). | 4 Days | Every SIG private email list |
| PMC Member Removal | When removal of a PMC member is sought. Note: Such actions will be triggered by TOC | 2/3 Majority | Active PMC members (excluding the member in question). | 4 Days | Every SIG private email list |
| Create a new SIG | Create a new SIG with initial PMCs and Members, Note: Such actions will be triggered by TOC | 2/3 Majority | TOC members | 4 Days | [https://github.com/oceanbase/oceanbase/discussions](https://github.com/oceanbase/oceanbase/discussions) |
| Modifying Bylaws | Modifying Bylaws | 2/3 Majority | TOC members | 4 Days | [https://github.com/oceanbase/oceanbase/discussions](https://github.com/oceanbase/oceanbase/discussions) |

**Comment by whhe** [comment]
Membership can also be managed without mailing lists, like k8s community:
https://github.com/kubernetes/community/blob/master/github-management/new-membership-procedure.md

Here new member is not invited by the community, the contributor should raise an apply himself for it.

**Reply by longdafeng** [reply]
It is ok for adding members in public,  but it is tough to remove member in public. 

We can run with email list for half a year, then we turn around in github discussion later.

**Comment by longdafeng** [comment]
This proposal has been proposed for 24 hours. 

Let's vote for this proposal. If this proposal has been past, it will be display in the official website.

**Comment by longdafeng** [comment]
It seems this discussion can't be changed from category of classroom to polls. 
The next time, we will use poll category for voting. 

+1(binding)

**Reply by Chuanhui** [reply]
+1（binding）

**Comment by xiangjiang-wn** [comment]
+1（binding）

**Comment by longdafeng** [comment]
Thanks to all. This proposal has been passed by all TOC members. 
The result is as following:
+1  -- 3 votes
+0 -- 0 vote
-0 -- 0 vote
-1 -- 0 vote


The OceanBase Community will be displayed in the official website .
